### PR TITLE
Better graphs for scatters using subgraphs (and implicit Monoids!)

### PR DESCRIPTION
**Pull request by cjllanwarne**
Also brings in the WDL4S cromwom merge.

Example using the centaur `ifs_in_scatters` test:
```
workflow ifs_in_scatters {
  Array[Int] numbers = range(5)

  scatter (n in numbers) {

    call validate_int { input: i = n }
    if (validate_int.validation) {
      Int incremented = n + 1
      call mirror { input: i = incremented }
    }
  }

  output {
    Array[Int?] mirrors = mirror.out
  }
}
```

Before:
graph

After:
graph

### CODE REVIEW

File: src/main/scala/wdltool/GraphPrint.scala [hunk 1]
```diff
@@ -1,71 +1,101 @@
...
-import wdl4s.{CallOutput, Declaration, If, Scatter, _}
 import scala.collection.JavaConverters._
 
+import cats._
+import cats.implicits._
+
 object GraphPrint {
 
-  case class WorkflowDigraph(workflowName: String, digraph: Set[String])
+  case class WorkflowDigraph(workflowName: String, digraph: NodesAndLinks)
```
Comment (geoffjentry) [inline src/main/scala/wdltool/GraphPrint.scala]: finals

File: src/main/scala/wdltool/GraphPrint.scala [hunk 2]
```diff
@@ -1,71 +1,101 @@
...
+  case class NodesAndLinks(nodes: Set[String], links: Set[String])
+  implicit val scopesAndLinksMonoid = new Monoid[NodesAndLinks] {
```
Comment (geoffjentry) [inline src/main/scala/wdltool/GraphPrint.scala]: `ScopesAndLinksMonoid`

File: src/main/scala/wdltool/GraphPrint.scala [hunk 3]
```diff
@@ -1,71 +1,101 @@
...
   }
 
-  private def defaultFilter: GraphNode => Boolean = _ => true
+  private def defaultFilter: WdlGraphNode => Boolean = _ => true
+
+  val clusterCount: AtomicInteger = new AtomicInteger(0)
+
 
-  private def listAllGraphNodes(namespace: WdlNamespaceWithWorkflow, filter: GraphNode => Boolean = defaultFilter): Set[String] = {
+  private def listAllGraphNodes(namespace: Scope): NodesAndLinks = {
```
Comment (geoffjentry) [inline src/main/scala/wdltool/GraphPrint.scala]: I think the answer is "yes" but is `namespace` still an appropriate name considering the type changed from a `Namespace` to a `Scope`? I think conceptually you could still view a scope as a namespace so it's fine, but throwing it out there

**Comment by geoffjentry** [comment]
Pretty cool. @vdauwera can you take a peek at this in terms of the visualization? I believe this was a request by @yfarjoun but he's on vaca

File: src/main/scala/wdltool/GraphPrint.scala (hunk 2 above)
Comment (mcovarr) [reply src/main/scala/wdltool/GraphPrint.scala]: It seems kind of tedious to have to write a dead obvious monoid instance for a case class whose two members already have monoid instances.  Is there not some magic to dispel this?

**Comment by mcovarr** [comment]
very nice!  :+1:

[Approved with PullApprove](https://pullapprove.com/broadinstitute/wdltool/pull-request/33/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=broadinstitute/wdltool)

File: src/main/scala/wdltool/GraphPrint.scala (hunk 2 above)
Comment (danbills) [reply src/main/scala/wdltool/GraphPrint.scala]: There is indeed: [kittens](https://github.com/milessabin/kittens)

Comment (mcovarr) [reply src/main/scala/wdltool/GraphPrint.scala]: 😸

Comment (danbills) [reply src/main/scala/wdltool/GraphPrint.scala]: I submitted a PR to derive the monoid: https://github.com/broadinstitute/wdltool/pull/34

Comment (mcovarr) [reply src/main/scala/wdltool/GraphPrint.scala]: kittens *are* magic

File: src/main/scala/wdltool/GraphPrint.scala (hunk 3 above)
Comment (cjllanwarne) [reply src/main/scala/wdltool/GraphPrint.scala]: I think the answer is probably no, it's used recursively and the inner scopes are definitely not namespaces

Comment (geoffjentry) [reply src/main/scala/wdltool/GraphPrint.scala]: I meant "namespace" in the colloquial sense, not an official wdl object. But really I was just giving you an easy rationalization to get out of changing things, so up to you :)

**Comment by gemmalam** [comment]
@geoffjentry looks like you said you would review this, but haven't quite given it a 👍

File: src/main/scala/wdltool/GraphPrint.scala [hunk 4]
```diff
@@ -1,71 +1,95 @@
...
+import cats.implicits._
+import cats.derived.monoid._, legacy._
+
 object GraphPrint {
 
-  case class WorkflowDigraph(workflowName: String, digraph: Set[String])
+  final case class WorkflowDigraph(workflowName: String, digraph: NodesAndLinks)
+  final case class NodesAndLinks(nodes: Set[String], links: Set[String])
+
+  sealed trait Link{
```
Comment (geoffjentry) [inline src/main/scala/wdltool/GraphPrint.scala]: `Link {`

**Comment by geoffjentry** [comment]
:+1: top notch dotting

[Approved with PullApprove](https://pullapprove.com/broadinstitute/wdltool/pull-request/33/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=broadinstitute/wdltool)

**Pull request merged**